src/notebooks_service.py

```diff
@@ -379,15 +380,39 @@ def launch_notebook(user, namespace, project, commit_sha, notebook=None):
             mimetype='application/json'
         )
 
-    # 1. launch using spawner that checks the access
+    ## 1. launch using spawner that checks the access
     headers = {auth.auth_header_name: 'token {0}'.format(auth.api_token)}
 
-    # if there is an image passed with the request, use it
+    # set the notebook image
     image = get_notebook_image(user, namespace, project, commit_sha)
 
-    default_url = request.args.get(
-        'default_url', os.environ.get('JUPYTERHUB_SINGLEUSER_DEFAULT_URL')
+    ## process the server options
+    # server options from system configuration
+    server_options_file = os.getenv(
+        'NOTEBOOKS_SERVER_OPTIONS_PATH',
+        '/etc/renku-notebooks/server_options.json'
+    )
+
+    with open(server_options_file) as f:
+        server_options_defaults = json.load(f)['serverOptions']
+    default_resources = server_options_defaults.get(
+        'resources', {}
     )
+    # process the requested options and set others to defaults from config
+    server_options = request.get_json().get('server_options', {})
+    server_options.setdefault(
+        'default_url',
+        server_options_defaults.get('defaultUrl', {}).get(
+            'default', os.getenv('JUPYTERHUB_SINGLEUSER_DEFAULT_URL')
+        )
+    )
+    server_options.setdefault('resources', {})
+
+    for key in default_resources.keys():
+        server_options['resources'].setdefault(
+            key,
+            default_resources.get(key)['default']
+        )
```

Comment:
TODO: validate options

Reply:
Yes, I will make an issue for this -- it's not obvious to me how the validation should happen or indeed what should be validated.